feature/unit-tests

MMurrayAtBR opened the pull request:
Creation of new  and updates to existing unit test classes for the app and data modules.

Commit:
Unit Tests

- Updated BaseTest with Before/After to set & reset Dispatchers.Main
- Moved Test Rules to BaseTest
- Clean up of HomeViewModelTest.kt. Moved the bitBucket mock out of the file to be used elsewhere.
- Created a MockBitBucketRepo object so the mock can be used across all app module test classes
- Added more tests to HomeViewModelTest.kt
- Changed "srcFiles" & "currentRepoName" to public to test; need to find a solution to access these values when private
- Added tests to RepositoryBrowserViewModelTest.kt

Commit:
Unit Tests

- Updated BaseTest with Before/After to set & reset Dispatchers.Main
- Moved Test Rules to BaseTest
- Clean up of HomeViewModelTest.kt. Moved the bitBucket mock out of the file to be used elsewhere.
- Created a MockBitBucketRepo object so the mock can be used across all app module test classes
- Added more tests to HomeViewModelTest.kt
- Changed "srcFiles" & "currentRepoName" to public to test; need to find a solution to access these values when private
- Added tests to RepositoryBrowserViewModelTest.kt
- Added Test classes, AuthCodeViewModelTest.kt, MockBitBucketRepo.kt, RepositoryBrowserViewModelTest.kt, RepositoryFileViewModelTest.kt, CreateSnippetViewModelTest.kt, ProfileViewModelTest.kt, SnippetsViewModelTest.kt
- Added MockBuildConfigProvider

Commit:
Unit Tests

- Updated BaseTest with Before/After to set & reset Dispatchers.Main
- Moved Test Rules to BaseTest
- Clean up of HomeViewModelTest.kt. Moved the bitBucket mock out of the file to be used elsewhere.
- Created a MockBitBucketRepo object so the mock can be used across all app module test classes
- Added more tests to HomeViewModelTest.kt
- Changed "srcFiles" & "currentRepoName" to public to test; need to find a solution to access these values when private
- Added tests to RepositoryBrowserViewModelTest.kt
- Added Test classes, AuthCodeViewModelTest.kt, MockBitBucketRepo.kt, RepositoryBrowserViewModelTest.kt, RepositoryFileViewModelTest.kt, CreateSnippetViewModelTest.kt, ProfileViewModelTest.kt, SnippetsViewModelTest.kt, SplashViewModelTest.kt
- Added MockBuildConfigProvider

Commit:
Unit Tests

- Updated BaseTest with Before/After to set & reset Dispatchers.Main
- Moved Test Rules to BaseTest
- Clean up of HomeViewModelTest.kt. Moved the bitBucket mock out of the file to be used elsewhere.
- Created a MockBitBucketRepo object so the mock can be used across all app module test classes
- Added more tests to HomeViewModelTest.kt
- Changed "srcFiles" & "currentRepoName" to public to test; need to find a solution to access these values when private
- Added tests for all ViewModel class in App Module
- Added MockBuildConfigProvider

Commit:
Unit Tests [updates]

- Updated date.model tests to 100%

Commit:
Unit Tests [updates]

- Added test classes for Commit, ComposeActivityViewModel, RepoFiles, RepositoryBrowserData, RepositoryFileData, Routes, SnippetDto, UserDtoConverter, and WorkspaceDto

- Updated DevOptionsViewModelTest.kt with a test for "manually" updating the environment spinner

- Updated RepositoryBrowserViewModelTest.kt with a new test for testing the path variable when path data is null

BottleRocket-Colin commented:
@MMurrayAtBR - This PR is currently pointing at develop.  Please change to merge into compose/develop.

BottleRocket-Colin commented on app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/repository/RepositoryBrowserViewModel.kt:
```diff
@@ -18,8 +18,8 @@ class RepositoryBrowserViewModel : BaseViewModel() {
-    private val srcFiles = MutableStateFlow<List<RepoFile>>(emptyList())
-    private var currentRepoName: String = ""
+    val srcFiles = MutableStateFlow<List<RepoFile>>(emptyList())
+    var currentRepoName: String = ""
```
In general, we don't want to make items public just for testing purpose.  Try to get same coverage by testing what these elements depend on and keeping these private.

MMurrayAtBR replied:
Will do!

jaredstockton replied:
@MMurrayAtBR I agree with @BottleRocket-Colin's original statement here that it would be best if these properties were left as private. I see that you've added the `@VisibleForTesting` annotation instead, which isn't ideal but should be okay if we use it sparsely.

Just keep in mind for future tests that it is always preferred to prevent making visibility changes just to make testing easier over changing the visibility and using the annotation :)

BottleRocket-Colin reviewed (changes requested):
One minor update.  Love the detailed mocks and clean test flows for state flows!!!  Good stuff.

jaredstockton commented on app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/splash/SplashViewModel.kt:
```diff
@@ -15,6 +15,10 @@ class SplashViewModel : BaseViewModel() {
+        authenticate()
+    }
+
+    fun authenticate() {
```
@MMurrayAtBR I see that you've created a public function just to test out this behavior. I'm currently exploring alternatives to test this with the code staying inside the `init` block. In the meantime, go ahead and add the `@VisibleForTesting` annotation to the `authenticate` function

MMurrayAtBR replied:
No longer applicable, so I'll resolve this conversation

jaredstockton committed:
Review commit: Run `ktlintFormat`

jaredstockton committed:
Review commit: Add `:compose - testInternalDebugUnitTest` Android Studio Run Configuration to run all tests within the compose module.

jaredstockton committed:
Review commit: Refactor tests to use updated coroutines testing APIs in 1.6.0+ (#71)

* Replace `runBlocking` with `runTest`.
* Overhaul the TestDispatcherProvider to use Unconfined implementation of a TestDispatcher.
* Allow each test class to override the TestDispatcherProvider used to run tests with a Standard implementation.
* Remove the the manual collection and collector cancellation from all tests, relying solely on Turbine's `test` functionality for hot/cold flows instead.
* Inline the SplashViewModel authenticate function back into the init block and use the StandardTestDispatcher to allow the event to be captured.

Commit:
Feature Unit Tests

- Updated RepositoryBrowserViewModel values srcFiles and currentRepoName with @VisibleForTesting annotation

MMurrayAtBR commented on app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/repository/RepositoryBrowserViewModel.kt:
```diff
@@ -18,8 +19,10 @@ class RepositoryBrowserViewModel : BaseViewModel() {
     // State
-    private val srcFiles = MutableStateFlow<List<RepoFile>>(emptyList())
-    private var currentRepoName: String = ""
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
```
I know that private is the default setting, but I add the otherwise value to prevent any confusion

jaredstockton reviewed (approved):
Thanks for all the work here!
